### PR TITLE
Memory manager support for Windows nodes

### PR DESCRIPTION
**Pull request by marosset**
/kind feature

#### What this PR does / why we need it:

This PR adds support for memory manager on Windows nodes and add a new `BestEffort` policy which allows giving preferences to the Windows OS for which CPUs to bind containers to.
Windows cannot guarantee scheduling workloads to specific NUMA nodes so instead we query the OS for which CPUs are part of a given NUMA node and specify that when starting containers.

#### Which issue(s) this PR fixes:

Part of https://github.com/kubernetes/kubernetes/issues/125262

To enable memory manager support the kubelet configuration would look something like

```toml
cpuManagerPolicy: static
memoryManagerPolicy: BestEffort
systemReserved: 
  cpu: 500m
  memory: 500Mi
featureGates:
  MemoryManager: true
  WindowsCPUAndMemoryAffinity: true

```
#### Does this PR introduce a user-facing change?

```release-note
Added Windows support for the node memory manager.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-windows/4885-windows-cpu-and-memory-affinity
```

/sig windows node
/milestone v1.32
/area kubelet
/assign @jsturtevant

### CODE REVIEW

File: pkg/kubelet/cm/container_manager_windows.go
Comment (ffromani) [inline pkg/kubelet/cm/container_manager_windows.go:106]: please rebase as we merged https://github.com/kubernetes/kubernetes/pull/128517

File: pkg/kubelet/cm/container_manager_windows.go
Comment (ffromani) [inline pkg/kubelet/cm/container_manager_windows.go:107]: you can now use the value returned on line 96 above, like the linux container manager does.

File: pkg/kubelet/cm/internal_container_lifecycle_windows.go
Comment (ffromani) [inline pkg/kubelet/cm/internal_container_lifecycle_windows.go:122]: can we use sets from `apimachinery/util` (or its most current replacement like perhaps `k8s.io/utils`)?

File: pkg/kubelet/cm/memorymanager/memory_manager.go
Comment (ffromani) [inline pkg/kubelet/cm/memorymanager/memory_manager.go:163]:
```
if runtime.GOOS != "windows" {
    return nil, fmt.Errorf("policy \"%s\" is not avaiable for platform \"%s\"", policyTypeBestEffort, runtime.GOOS)
}
```
?

File: pkg/kubelet/cm/memorymanager/policy_best_effort.go
Comment (ffromani) [inline pkg/kubelet/cm/memorymanager/policy_best_effort.go:30]: interface?

File: pkg/kubelet/cm/memorymanager/policy_best_effort.go
Comment (ffromani) [inline pkg/kubelet/cm/memorymanager/policy_best_effort.go:32]: Is not obvious to me what do we gain by wrapping the static policy. Could you please elaborate a bit?

**Comment by ffromani** [comment]
/triage accepted
/priority important-soon

initial priority assessment subject to be changed

File: pkg/kubelet/cm/internal_container_lifecycle_windows.go
Comment (jsturtevant) [inline pkg/kubelet/cm/internal_container_lifecycle_windows.go:155]: This looks pretty close to https://github.com/kubernetes/kubernetes/blob/3140807126c86163584974e8a0835c2f5907af97/pkg/kubelet/winstats/cpu_topology.go#L91  

Could we provide another helper function and/or adjust the logic to convert to affinities?  My concern would be spreading the logic for calculating back and forth between kubelet's representation  (integers for the CPU's) to Windows representation of groupAffinities.

File: pkg/kubelet/cm/internal_container_lifecycle_windows.go
Comment (jsturtevant) [inline pkg/kubelet/cm/internal_container_lifecycle_windows.go:56]: Thanks for all the unit tests on the merging of sets.  There is quite a bit of logic here, that might be nice to have unit test to verify all this behavior (the three scenarios and with each manager on/off)

File: pkg/kubelet/cm/internal_container_lifecycle_windows.go
Comment (marosset) [reply pkg/kubelet/cm/internal_container_lifecycle_windows.go:122]: yes, i'll make some updates

File: pkg/kubelet/cm/memorymanager/memory_manager.go
Comment (marosset) [reply pkg/kubelet/cm/memorymanager/memory_manager.go:163]: oops

**Comment by ffromani** [comment]
@marosset adding the policy name ultimates surfaces to the user and to the kubeletconfig API object so this change *MAY* review api-review, please take this into account.

File: pkg/kubelet/cm/memorymanager/policy_best_effort.go
Comment (marosset) [reply pkg/kubelet/cm/memorymanager/policy_best_effort.go:32]: For Windows we want the new `BestEffort` policy to have the same logic as the existing `static` policy.
I figured that doing this would reduce a lot of code duplication.

I can update the comments here to reflect that or maybe I can have the kubelet arguments specify `BestEffortPolicy` and then create a `staticPolicy` instance here for Windows.

Do you all have a preference @ffromani @jsturtevant ?

Comment (jsturtevant) [reply pkg/kubelet/cm/memorymanager/policy_best_effort.go:32]: Updating comments and wrapping seems to make sense to me. This will allow to adjust more easily if we learn of reasons to adjust.

Comment (ffromani) [reply pkg/kubelet/cm/memorymanager/policy_best_effort.go:32]: I'll need to refresh my memory over what we discussed in the KEP before I can answer. Right now I can surely say comments is a great starting point.

File: pkg/kubelet/cm/internal_container_lifecycle_windows.go
Comment (marosset) [reply pkg/kubelet/cm/internal_container_lifecycle_windows.go:56]: agree, i'll try and add some

File: pkg/kubelet/cm/internal_container_lifecycle_windows.go
Comment (marosset) [reply pkg/kubelet/cm/internal_container_lifecycle_windows.go]: Switched to apimachinery/util since it had everything I needed.

File: pkg/kubelet/cm/internal_container_lifecycle_windows.go
Comment (marosset) [reply pkg/kubelet/cm/internal_container_lifecycle_windows.go]: I refactored some of the logic and got rid of the need for this method

File: pkg/kubelet/cm/internal_container_lifecycle_windows.go
Comment (marosset) [reply pkg/kubelet/cm/internal_container_lifecycle_windows.go]: I did some slight refactoring and added unit tests for this logic - PTAL

**Comment by sftim** [comment]
Changelog suggestion
```diff
-Windows: Support memory manager on Windows 
+Added Windows support for the node memory manager.
```

File: pkg/kubelet/cm/internal_container_lifecycle_windows.go
Comment (ffromani) [inline pkg/kubelet/cm/internal_container_lifecycle_windows.go:40]: you may want to set this to `V(2)` at least, likely `V(4)` otherwise it can get spammy very fast

File: pkg/kubelet/cm/memorymanager/memory_manager.go
Comment (ffromani) [inline pkg/kubelet/cm/memorymanager/memory_manager.go:160]: nit: you can do like `Static` above:
```
if runtime.GOOS != "windows" {
    return nil, fmt.Errorf("policy \"%s\" is not available for platform \"%s\"", policyTypeBestEffort, runtime.GOOS)
}
systemReserved, err := ...
```

File: pkg/kubelet/cm/memorymanager/memory_manager.go
Comment (ffromani) [inline pkg/kubelet/cm/memorymanager/memory_manager.go:145]: nit: can we use `%q` instead of `\"%s\"` ?

File: pkg/kubelet/cm/memorymanager/policy_best_effort.go
Comment (ffromani) [reply pkg/kubelet/cm/memorymanager/policy_best_effort.go:37]: ok, first: I've nothing against reusing the code like this. We didn't do often in the codebase, but why not?
The issue I see however is that from my interpretation of the KEP I would have expected different hint generation on windows based on https://github.com/kubernetes/enhancements/tree/master/keps/sig-windows/4885-windows-cpu-and-memory-affinity#windows-memory-considerations

I think we can settle this with a good set of testcases, at very least unit tests. which seems lacking to me?
Granted, the policy is actually a thin wrapper over the static policy, which has its own tests; the tests I recommend to add are to ensure and document the expected behavior on windows.

The tests should cover preferably also the topology manager merge hint process. It will be similar in spirit to what you are adding to `internal_container_lifecycle_windows_test.go`, but covering the generic topology manager flow.

There's likely not enough time to do this for this cycle. Plus I want to be practical: it's an alpha feature, disabled by default, and the early feedback from having some functionality in is important as well. So I don't think it is blocking, but I _DO_ think this is **beta**-blocking (to discuss this topic in detail at least) and possibly we can need another alpha iteration to sort this out depending on the aforementioned feedback

Comment (jsturtevant) [reply pkg/kubelet/cm/memorymanager/policy_best_effort.go:37]: > The issue I see however is that from my interpretation of the KEP I would have expected different hint generation on windows based on https://github.com/kubernetes/enhancements/tree/master/keps/sig-windows/4885-windows-cpu-and-memory-affinity#windows-memory-considerations

I didn't see a difference in what we would provide for a hint. The major difference is in the semantic meaning of the hint, which on Windows we can't guarantee that it the hint will be respected, only make that suggestion.  What kind of difference are you thinking?

>There's likely not enough time to do this for this cycle. Plus I want to be practical: it's an alpha feature, disabled by default, and the early feedback from having some functionality in is important as well. So I don't think it is blocking, but I DO think this is beta-blocking (to discuss this topic in detail at least) and possibly we can need another alpha iteration to sort this out depending on the aforementioned feedback

I appreciate this sentiment.  We do have a user we are working with that is willing to provide feedback but is looking for it to be in an Alpha release.  From my understanding of the challenges on Linux, these types of features need some real feedback to get them dialed in, and I anticipate we will get feedback that we will need to incorporate back into the design.

Comment (ffromani) [reply pkg/kubelet/cm/memorymanager/policy_best_effort.go:37]: > > The issue I see however is that from my interpretation of the KEP I would have expected different hint generation on windows based on https://github.com/kubernetes/enhancements/tree/master/keps/sig-windows/4885-windows-cpu-and-memory-affinity#windows-memory-considerations
> 
> I didn't see a difference in what we would provide for a hint. The major difference is in the semantic meaning of the hint, which on Windows we can't guarantee that it the hint will be respected, only make that suggestion. What kind of difference are you thinking?

I'm basing my expectations on what I saw so far on linux for the cpumanager policy. I would like to see if for example returning a different set of NUMA bitmasks or different values for `Preferred` field.

If we add a thin wrapper over the Static policy, we are saying that this policy should behave 1:1 like the static policy, and we are unnecessarily coupling the behavior. What I would rather see is a specification (and unit tests are a good vehicle for that) of the expected behavior, more precise than (yes, I'm oversimplifying here) "whatever Static policy does, but not enforcing".

Lacking specification is something that biten us already multiple already, so we should IMO go in the direction to have the spec spelled out more precisely, and the testcase well defined. If, after that, we can reuse the implementation, great.

If nothing else (which I can accept!) this should be discussed and elaborated in the KEP design details.
I'm happy to elaborate further in the next cycle. Again, more details in the KEP to elaborate more precisely the behavior can possibly be all we need.

> > There's likely not enough time to do this for this cycle. Plus I want to be practical: it's an alpha feature, disabled by default, and the early feedback from having some functionality in is important as well. So I don't think it is blocking, but I DO think this is beta-blocking (to discuss this topic in detail at least) and possibly we can need another alpha iteration to sort this out depending on the aforementioned feedback
> 
> I appreciate this sentiment. We do have a user we are working with that is willing to provide feedback but is looking for it to be in an Alpha release. From my understanding of the challenges on Linux, these types of features need some real feedback to get them dialed in, and I anticipate we will get feedback that we will need to incorporate back into the design.

Fine, so we will evaluate during the 1.33 cycle if we can move to beta (taking into account my oen feedback above) or if we need another iteration.

The rest of the code overall LGTM, I'll have another pass ASAP but from my recollections I don't have anything major on my notes.

File: pkg/kubelet/cm/memorymanager/policy_best_effort.go
Comment (jsturtevant) [inline pkg/kubelet/cm/memorymanager/policy_best_effort.go:29]:
```suggestion
// but unlike linux based systems, on Windows systems numa nodes cannot be directly assigned or guaranteed via Windows APIs
```

File: pkg/kubelet/cm/memorymanager/policy_best_effort.go
Comment (jsturtevant) [inline pkg/kubelet/cm/memorymanager/policy_best_effort.go:30]:
```suggestion
// (windows scheduler will use the numa node that is closest to the cpu assigned therefor respecting the numa node assignment as a best effort). Because of this we don't want to have users specify "StaticPolicy" for the memory manager
```

or something like that?

File: pkg/kubelet/cm/internal_container_lifecycle_windows.go
Comment (jsturtevant) [reply pkg/kubelet/cm/internal_container_lifecycle_windows.go]: Thanks!

**Review by ffromani — COMMENTED** [review]
LGTM once comments have been addressed.

I think we should *try* to reconcile in the next cycle the windows-specific alignment code with the generic code from topology manager. IOW, instead of asking cpumanager for the affinities, we should try to query topology manager, which is supposed to be the source of truth for alignment.

TL;DR: the windows-specific code is reimplementing a narrowed-down version of topology manager, and we should try to converge back in topology manager in the future.

[EDIT] I'm aware the linux container internal container lifecycle code does the same; but it's also true that up until now that was the only true implementation. Now that we have two, we should give an honest attempt to generalize. Worst case we keep the _status quo_ with comments explaining why.

**Comment by marosset** [comment]
> Changelog suggestion
> 
> ```diff
> -Windows: Support memory manager on Windows 
> +Added Windows support for the node memory manager.
> ```

Done

File: pkg/kubelet/cm/internal_container_lifecycle_windows.go
Comment (marosset) [reply pkg/kubelet/cm/internal_container_lifecycle_windows.go]: updated to v(4)

**Comment by marosset** [comment]
thanks @ffromani 
I think I addressed all your feedback - please take another look when you get a chance

I'll open up a separate PR to include unit testing for the BestEffort policy as a `beta` graduation requirement to the KEP.

**Comment by marosset** [comment]
> @marosset adding the policy name ultimates surfaces to the user and to the kubeletconfig API object so this change _MAY_ require api-review, please take this into account.

@ffromani - Do you know who would be able to make that decision?

**Comment by marosset** [comment]
> > @marosset adding the policy name ultimates surfaces to the user and to the kubeletconfig API object so this change _MAY_ require api-review, please take this into account.
> 
> @ffromani - Do you know who would be able to make that decision?

/label api-review
In case this does...

**Comment by marosset** [comment]
It looks like there are some unrelated infra issue with the windows / CAPZ test passes - please ignore those failure for now.

**Comment by jsturtevant** [comment]
We are still recovering from an issue with the azure sub
/test pull-kubernetes-e2e-capz-windows-master

**Comment by jsturtevant** [comment]
I've pulled this branch and run a few different scenarios, and I am seeing the expected behavior.  A few of the new unit tests did fail but they looked like typos nothing serious.  Once those are resolved this lgtm.

**Comment by marosset** [comment]
> I've pulled this branch and run a few different scenarios, and I am seeing the expected behavior. A few of the new unit tests did fail but they looked like typos nothing serious. Once those are resolved this lgtm.

The unit test caught an issue that got introduced by a small refactor I made a few commits ago
https://github.com/kubernetes/kubernetes/pull/128560/files#diff-6711ce20b4cecd0fa883ddc3f1c6dde5c56dcfebca347b4de3e6e794feab5269R111 was incorrect.

All the unit tests are passing on Windows now.
Thanks for pointing this out @jsturtevant

**Review by ffromani — COMMENTED** [review]
/lgtm

**Comment by ffromani** [comment]
@marosset / @jsturtevant is not clear if this PR is _needed_ for https://github.com/kubernetes/enhancements/issues/4885 (I'm fine in both cases per my comments above). If so, can you please update the list on the issue? Thanks!

**Comment by ffromani** [comment]
folks kindly be aware of https://github.com/kubernetes/kubernetes/pull/128657

**Comment by ffromani** [comment]
/lgtm cancel

need to rebase on top of https://github.com/kubernetes/kubernetes/pull/128657

**Comment by marosset** [comment]
> /lgtm cancel
> 
> need to rebase on top of #128657

Ack - I'll rebase on that as soon as it merges.

**Review by ffromani — COMMENTED** [review]
/lgtm

thanks for the rebase and sorry for the churn. You use the recently added `containerMap.Clone()`, so looks fine again to me.

File: pkg/kubelet/cm/memorymanager/memory_manager.go
Comment (mrunalp) [inline pkg/kubelet/cm/memorymanager/memory_manager.go:145]: nit: typo: `not not`

**Review by mrunalp — APPROVED** [review]

**Comment by ffromani** [comment]
beta-blocking (or perhaps as later bugfix?) we need to update `cmd/kubelet/app/options/options.go`

File: pkg/kubelet/cm/memorymanager/memory_manager.go
Comment (marosset) [reply pkg/kubelet/cm/memorymanager/memory_manager.go]: fixed

**Comment by tengqm** [comment]
Do we have documentation updates (including the feature gate) to the website?

**Comment by marosset** [comment]
> Do we have documentation updates (including the feature gate) to the website?

I we'll use https://github.com/kubernetes/website/pull/48469 to update docs for both cpu manager (https://github.com/kubernetes/kubernetes/pull/125296) and memory manager (this PR) support for windows.

**Pull request merged**
